pkg/knowledge_base/edge_kb.go

```diff
@@ -221,74 +226,72 @@ func (kb EdgeKB) ExpandEdges(dag *core.ResourceGraph, appName string) (err error
 
 		zap.S().Debugf("Found valid paths %s", validPaths)
 		var validPath []Edge
-
-		// If we have more than 1 valid path we will always default to the direct path if it exists, otherwise we will raise an error since we cannot determine which path we are supposed to use.
-		if len(validPaths) > 1 {
-			for _, p := range validPaths {
-				if len(p) == 1 {
-					zap.S().Debug("Defaulting to direct path")
-					validPath = p
-				}
-			}
+		// Get the shortest route that satisfied constraints
+		for _, path := range validPaths {
 			if len(validPath) == 0 {
-				merr.Append(fmt.Errorf("found multiple paths which satisfy constraints for edge %s -> %s. \n Paths: %s", dep.Source.Id(), dep.Destination.Id(), validPaths))
+				validPath = path
+			} else if len(path) < len(validPath) {
+				validPath = path
+			} else if len(path) == len(validPath) {
+				merr.Append(fmt.Errorf("found multiple paths which satisfy constraints for edge %s -> %s and are the same length. \n Paths: %s", dep.Source.Id(), dep.Destination.Id(), validPaths))
```

Comment:
nit: would like both paths to be in the error so you can see which two collide (the current error only has one path)

Reply:
so it does log validPaths, i think that should be covered right?

Reply:
Oh, it logs all the valid paths - I missed the `s`. IMO just put the two conflicting ones, but all is fine too.